### PR TITLE
Disinvest revised

### PR DESCRIPTION
**Pull request by 0xrin1**
- Reworked from https://github.com/lindy-labs/sc_solidity-contracts/pull/48/files
- Adds disinvest functionality to updateInvested function
- Introduces disinvest event
- Introduces synchronous mock disinvest for testing
- Introduces investState function to determine whether investment or disinvestment is needed
- Corresponding tests

### CODE REVIEW

File: contracts/strategy/anchor/AnchorStrategy.sol [hunk 1]
```diff
@@ -145,9 +145,17 @@ contract AnchorStrategy is IStrategy, AccessControl {
      */
     function withdrawToVault(uint256 amount)
         external
+        virtual
         override(IStrategy)
         onlyManager
-    {}
+    {
+        require(amount > 0, "AnchorStrategy: amount is zero");
```
Comment (coolhill) [inline contracts/strategy/anchor/AnchorStrategy.sol:152]: can use `!= 0` for gas savings

File: test/Vault.spec.ts [hunk 2]
```diff
@@ -501,43 +501,75 @@ describe("Vault", () => {
...
+
+        const tx = await vault.connect(owner).updateInvested();
+
+        await expect(tx).to.emit(vault, "Disinvested").withArgs(parseUnits("10"));
+      });
     });
   });
 
+  // TODO: change
   describe("investableAmount", () => {
```
Comment (joaop21) [inline test/Vault.spec.ts:561]: The function name is not `investableAmount` anymore. Is it ok to have a testing block with this name? Or might be confusing? Should we merge this tests in the next testing block?

**Review by joaop21 — COMMENTED** [review]
LGTM! 👍

File: test/Vault.spec.ts (hunk 2 above)
Comment (0xrin1) [reply test/Vault.spec.ts:561]: good shout overlooked this :see_no_evil:

Comment (0xrin1) [reply test/Vault.spec.ts:561]: merged it in with the `investState` test case describe block since they're testing the same core functionality

File: contracts/Vault.sol [hunk 3]
```diff
@@ -331,15 +333,32 @@ contract Vault is
     {
         require(address(strategy) != address(0), "Vault: strategy is not set");
 
-        uint256 _investable = investableAmount();
+        (uint256 maxInvestableAmount, uint256 alreadyInvested) = investState();
+
+        require(
+            maxInvestableAmount > alreadyInvested
+                ? maxInvestableAmount - alreadyInvested != 0
+                : alreadyInvested - maxInvestableAmount != 0,
```
Comment (gabrielpoca) [inline contracts/Vault.sol:341]: Maybe I'm confused, but isn't this the same as 

```
require(maxInvestableAmount != alreadyInvested, "Vault: nothing to invest or disinvest");
```

?

File: contracts/Vault.sol [hunk 4]
```diff
@@ -311,16 +311,18 @@ contract Vault is
...
 
-    function investableAmount() public view returns (uint256) {
-        uint256 maxInvestableAssets = totalUnderlying().pctOf(investPct);
-
-        uint256 alreadyInvested = strategy.investedAssets();
-
-        if (alreadyInvested >= maxInvestableAssets) {
-            return 0;
+    function investState()
+        public
```
Comment (gabrielpoca) [inline contracts/Vault.sol:315]: Do we maybe want to make this external so the backend can quickly check before calling the updateInvested ?

File: contracts/strategy/anchor/AnchorStrategy.sol [hunk 5]
```diff
@@ -189,6 +197,11 @@ contract AnchorStrategy is IStrategy, AccessControl {
         );
     }
 
+    // @inheritdoc IStrategy
+    function disinvest() external virtual override(IStrategy) onlyManager {
```
Comment (gabrielpoca) [inline contracts/strategy/anchor/AnchorStrategy.sol:201]: What's this for?

File: test/Vault.spec.ts [hunk 6]
```diff
@@ -501,61 +501,116 @@ describe("Vault", () => {
 
     it("reverts if no investable amount", async () => {
       await vault.connect(owner).setStrategy(strategy.address);
-      await addYieldToVault("10");
```
Comment (gabrielpoca) [inline test/Vault.spec.ts:504]: Why did you take it out? Shouldn't this test still work without changes?

File: contracts/Vault.sol (hunk 3 above)
Comment (0xrin1) [reply contracts/Vault.sol]: yes :rofl: rofl: rofl: 

adjusted, sorry about that, I blame sleep deprivation

File: contracts/strategy/anchor/AnchorStrategy.sol (hunk 5 above)
Comment (0xrin1) [reply contracts/strategy/anchor/AnchorStrategy.sol:201]: this part is WIP from this morning, realized we were missing a InitRedeemDeposit call in the disinvest case of the new updateInvested function and i figured i'd make a new function to put the logic in there

like that we have an invest and disinvest function, not just an invest function

File: contracts/Vault.sol (hunk 4 above)
Comment (0xrin1) [reply contracts/Vault.sol:315]: weird actually I just tried switching the public to external and I get an error

image

we should be able to call this from the backend regardless though, public makes it available to call off chain

File: test/Vault.spec.ts (hunk 6 above)
Comment (0xrin1) [reply test/Vault.spec.ts:504]: in the old version of this PR  (`disinvest` branch) this still worked because in the test the `invest` function was called directly

when we use updateInvested and run addYieldToVault('10') beforehand, the updateInvested function disinvests instead of investing so the assertion at the bottom of the test fails

File: contracts/strategy/anchor/AnchorStrategy.sol (hunk 5 above)
Comment (0xrin1) [reply contracts/strategy/anchor/AnchorStrategy.sol:201]: nevermind, the withdrawToVault function does exactly this, removed this

File: contracts/Vault.sol (hunk 4 above)
Comment (naps62) [reply contracts/Vault.sol:315]: I don't understand. being public makes it external as well. and if it's used internally (according to that screenshot), then `public` is the right one

File: contracts/Vault.sol [hunk 7]
```diff
@@ -311,16 +311,18 @@ contract Vault is
...
+        view
+        override
```
Comment (naps62) [inline contracts/Vault.sol:317]: can we set `override(IVault)` to be more explicit?

File: contracts/Vault.sol (hunk 4 above)
Comment (gabrielpoca) [reply contracts/Vault.sol:316]: sorry, I was confused, you guys are right

File: contracts/Vault.sol [hunk 8]
```diff
@@ -331,15 +334,30 @@ contract Vault is
     {
         require(address(strategy) != address(0), "Vault: strategy is not set");
 
-        uint256 _investable = investableAmount();
+        (uint256 maxInvestableAmount, uint256 alreadyInvested) = investState();
+
+        require(
+            maxInvestableAmount != alreadyInvested,
+            "Vault: nothing to invest or disinvest"
```
Comment (gabrielpoca) [inline contracts/Vault.sol:341]: Can we shorten this to save some bytecode? `Vault: nothing to do` for instance

**Review by gabrielpoca — APPROVED** [review]
